### ISSUE TITLE
Severe Training Slowdown – xFormers Ignored, SDPA Used Instead

### ISSUE THREAD

**Issue by WackyArt** [bmaltais/kohya_ss#3200]
Hi there,

I’ve run into a major performance issue after switching to PyTorch ≥ 2.2 while training SDXL LoRA models with Kohya (using Prodigy optimizer and 2048px images).

Before: ~4–5h training time for 1800 steps (60 images, batch 1)
Now: 17h+, despite xFormers being selected in the GUI.

🧪 Observations:
Console shows:

torch.autograd.graph.py:825: UserWarning: cuDNN SDPA backward got grad_output.strides() != output.strides() ...

Training clearly uses cuDNN SDPA, not xFormers.

SDPA is causing massive overhead on my system (Windows 11, RTX 4060 Ti, 16GB VRAM).

❗ Expected Behavior:
If xFormers is selected in the GUI, it should be strictly prioritized.
Fallbacks to SDPA or other attention backends should be opt-in only, or at least clearly communicated in logs/UI.

✅ Request:
Please add a force-disable for SDPA and/or a clear toggle between attention backends.

Clarify if PyTorch >=2.2 auto-enables SDPA even when xFormers is active, and how to prevent that.

This change would help avoid 10+ hour training delays and ensure users get the expected performance out of their hardware.

Thanks for your awesome work – just want to help others avoid this trap!

**Comment by Haoming02** [comment]
If you updated to PyTorch 2.7.0, it doesn't support xformers yet

**Comment by WackyArt** [comment]
Thanks for your reply! Just to clarify – I'm not using PyTorch 2.7.0, but rather PyTorch 2.1.2+cu121, which should still support xFormers properly.

My issue isn't related to 2.7 – instead, the problem is that SDPA seems to be used during training even though xFormers is selected in the Kohya GUI, and no SDPA-related arguments are passed. This results in a significant slowdown (e.g. 17h instead of ~5h for the same config).

Could it be that SDPA is being prioritized internally, even when xFormers is available?

Thanks again for looking into this!

**Comment by bmaltais** [comment]
Can you share the json config for the training? I will try to reproduce it at my end.

**Comment by WackyArt** [comment]
Hi! I've isolated a performance issue that occurs when memory_efficient_attention is enabled and gradient_checkpointing is disabled — even though xformers is selected in the config.

In this setup, training becomes extremely slow (from ~3h to over 130h total).
At every step, the following warning appears:

vbnet
Kopieren
Bearbeiten
UserWarning: cuDNN SDPA backward got grad_output.strides() != output.strides(), attempting to materialize a grad_output with matching strides...
This warning originates from PyTorch's cuDNN SDPA implementation (aten/native/cudnn/MHA.cpp)
and does not appear when xformers is working correctly — suggesting that SDPA is used internally in this setup.

To demonstrate the difference, I've attached two JSON configs with nearly identical settings:

✅ p3ach_2048px_prodigy.json: fast training (~6.6s/it)
❌ p3ach_2048px_prodigy_slow.json: same config, but with mem_eff_attn = true and gradient_checkpointing = false → leads to massive slowdown and the SDPA warning.

Thanks for checking it out! Let me know if logs or environment details are helpful.

[p3ach_2048px_prodigy_slow.json](https://github.com/user-attachments/files/20064592/p3ach_2048px_prodigy_slow.json)

[p3ach_2048px_prodigy.json](https://github.com/user-attachments/files/20064596/p3ach_2048px_prodigy.json)

**Comment by b-fission** [comment]
A few random observations and questions.
- Can you provide the exact version numbers of kohya_ss and pytorch that you're running?
- Are you keeping on eye on your vram usage while training? If it exceeds your GPU's capacity, training will slow down massively. Training at 2048 resolution with Gradient checkpointing=Off will require a lot of vram (around 33gb from what I'm seeing). 
- Gradient checkpointing is a good way to reduce vram usage (in exchange for slower training). That would explain why it seems to help on your config. But exceeding vram capacity will still severely impact performance.
- Memory efficient attention doesn't really save memory, and is much slower than xformers or sdpa. Would not recommend.
- The attention methods are chosen in the following order: Memory efficient attention (checkbox on), xformers (dropdown), sdpa (dropdown). If you wanted xformers and it failed to load, the training would fail at startup.
- The training log output will tell you which attention method is being used. In the screenshot below, I'm using xformers (while memory efficient attention is turned off)
![Image](https://github.com/user-attachments/assets/62cc87d1-7cce-4c74-942e-e655a098fecd)
- Do you *need* to train at 2048 resolution instead of 1024?

**Comment by WackyArt** [comment]
Thanks a lot for your help, and sorry for the late reply –
I briefly tried OneTrainer in between, but at this point… I’m starting to think I am the issue 😅

Can you provide the exact version numbers of kohya_ss and pytorch that you're running?
kohya_ss fork: bmaltais, commit f3e1cb8 (April 22, 2025)
PyTorch: 2.5.0+cu124
xformers: 0.0.28.post2
CUDA runtime: 12.4
Running on Windows 11 in Python 3.10 venv

Yes, I’m watching VRAM usage closely.
I’ve changed training to 1024×1536.
Despite using fp16, xformers, and gradient checkpointing, VRAM fills up and step time jumps to 40–60s.
Looks like PyTorch 2.5.0 on Windows still runs SDPA internally – even with TORCH_DISABLE_SDPA=1 set.
This might trigger RAM fallback during backward. Still trying to find a clean way to fully disable SDPA.

Yes, gradient checkpointing is enabled again.
I only turned it off temporarily to test performance impact – VRAM usage spiked and training slowed down even more. With it enabled, things are more stable, even if a bit slower.

I agree – memory efficient attention is definitely slower.
I only enabled it briefly to test if it would prevent SDPA from being triggered. Just trying to trace exactly where the fallback happens.

That’s really helpful, thank you.
xformers is loading correctly (confirmed in logs), so that path works.
But despite optimized settings for my 4060 Ti (fp16, xformers, checkpointing, 1024×1536), VRAM still fills up during training – and I don’t understand why.
Once full, it seems SDPA kicks in for some components and triggers fallback to RAM.
Still trying to pinpoint what’s causing the overload – maybe it’s one of the text encoders?

Confirmed – my logs show: “Enable xformers for U-Net”, so it’s definitely active.
Memory efficient attention is off.
Still getting the SDPA warning during backward, though – possibly from another part of the model?

I initially tried training at 2048,
but quickly realized it was too much for my setup and switched to 1024×1536 and 1536×1024.
That should be more manageable with batch size 1 and AdamW8bit on a 16 GB card –
but maybe there's still something I'm overlooking!?

Thanks again for taking the time to share your thoughts and guidance – I really appreciate it 🙏
Sorry for the late reply – I briefly looked into OneTrainer… but at this point, I’m starting to think I might just be the issue 😅

![Image](https://github.com/user-attachments/assets/4c9854de-22ae-418b-a939-b2fe10f30261)

**Comment by WackyArt** [comment]
At this point, I think that despite the optimizations (fp16, xformers, checkpointing, batch size 1),
my VRAM is still filling up – and once that happens, PyTorch falls back to system RAM.
That might be what triggers the SDPA backward path and the warning during training.

That means my GPU is the issue :)

**Comment by b-fission** [comment]
Okay, let's double-check. You're training a LoRA, right? Not accidentally landing on the **Dreambooth** tab? Your screenshot above has the message `Disable Diffusers' xformers` which should never appear when training a LoRA.

When I tried Dreambooth training with text encoders enabled, I do see the SDPA warnings. Updating to pytorch 2.6.0 got rid of those messages, but I can't say for sure if that fixes the underlying issue.

**Comment by WackyArt** [comment]
You're absolutely right — I mixed things up here 😅

Initially, I encountered this issue during LoRA training, which frustrated me enough to give OneTrainer a try. There, everything worked fine, and I was able to train a LoRA without problems.
Recently though, I described a Dreambooth issue — and forgot to clarify that I had switched contexts. Sorry for the confusion!

Back to the LoRA problem:
I ran into this slowdown with 2048px resolution, which previously worked without issues. Unfortunately, I don't remember which PyTorch version I was using back then.

As for the current Dreambooth behavior, my research suggests it's expected — likely due to a VRAM-to-RAM fallback because my GPU isn't strong enough.
But for LoRA, this shouldn't happen — especially since it worked fine before.

I'll now test with a newer PyTorch version to see if it resolves the LoRA slowdown.

Thanks again for your time and help — much appreciated! 🙏

**Comment by WackyArt** [comment]
ehm, I'm testing with runpod now (A6000) even there:

2025-05-18 09:53:06 INFO     epoch is incremented. current_epoch: 0, epoch: 1                                         train_util.py:779
/workspace/kohya_ss/venv/lib/python3.10/site-packages/torch/autograd/graph.py:825: UserWarning: cuDNN SDPA backward got grad_output.strides() != output.strides(), attempting to materialize a grad_output with matching strides... (Triggered internally at ../aten/src/ATen/native/cudnn/MHA.cpp:674.)
  return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
steps:   8%|██████                                                                  | 212/2544 [03:49<42:08,  1.08s/it, avr_loss=0.101]
epoch 2/12
2025-05-18 09:56:55 INFO     epoch is incremented. current_epoch: 1, epoch: 2                                         train_util.py:779
steps:  14%|██████████                                                             | 361/2544 [06:30<39:19,  1.08s/it, avr_loss=0.0868]

Strange, I've thought this issues comes from too less vram ...

**Comment by b-fission** [comment]
From what I can tell, xformers is only applicable to the model's unet. The text encoder(s) might be triggering that `cuDNN SDPA` message on pytorch 2.5.0 regardless of the chosen CrossAttention. Have you tried a newer pytorch on RunPod?

**Comment by WackyArt** [comment]
No, I haven’t updated PyTorch yet – I assumed the SDPA fallback warning was only related to low VRAM conditions. But as you can see, the same issue happens on a RunPod A6000, and the GPU memory isn't even fully used. Since training still runs at around 1.08 it/s, I’m okay with the warning for now. Honestly, I’m a bit hesitant to touch my environment if everything else works – I’d rather accept the warning than risk breaking something. 😅
